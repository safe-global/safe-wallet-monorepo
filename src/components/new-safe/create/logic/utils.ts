import { isSmartContract } from '@/utils/wallets'
import { type ChainInfo } from '@safe-global/safe-gateway-typescript-sdk'
import { sameAddress } from '@/utils/addresses'
import { createWeb3ReadOnly, getRpcServiceUrl } from '@/hooks/wallets/web3'
import { type ReplayedSafeProps } from '@/store/slices'
import { predictAddressBasedOnReplayData } from '@/components/welcome/MyAccounts/utils/multiChainSafe'

export const getAvailableSaltNonce = async (
  customRpcs: {
    [chainId: string]: string
  },
  replayedSafe: ReplayedSafeProps,
  chains: ChainInfo[],
  // All addresses from the sidebar disregarding the chain. This is an optimization to reduce RPC calls
  knownSafeAddresses: string[],
): Promise<string> => {
  let isAvailableOnAllChains = true
  const allRPCs = chains.map((chain) => {
    const rpcUrl = customRpcs?.[chain.chainId] || getRpcServiceUrl(chain.rpcUri)
    // Turn into Eip1993Provider
    return {
      rpcUrl,
      chainId: chain.chainId,
    }
  })

  for (const chain of chains) {
    const rpcUrl = allRPCs.find((rpc) => chain.chainId === rpc.chainId)?.rpcUrl
    if (!rpcUrl) {
      throw new Error(`No RPC available for  ${chain.chainName}`)
    }
    const web3ReadOnly = createWeb3ReadOnly(chain, rpcUrl)
    if (!web3ReadOnly) {
      throw new Error('Could not initiate RPC')
    }
    const safeAddress = await predictAddressBasedOnReplayData(replayedSafe, web3ReadOnly)
    const isKnown = knownSafeAddresses.some((knownAddress) => sameAddress(knownAddress, safeAddress))
    if (isKnown || (await isSmartContract(safeAddress, web3ReadOnly))) {
      // We found a chain where the nonce is used up
      isAvailableOnAllChains = false
      break
    }
  }

  // Safe is already deployed so we try the next saltNonce
  if (!isAvailableOnAllChains) {
    return getAvailableSaltNonce(
      customRpcs,
      { ...replayedSafe, saltNonce: (Number(replayedSafe.saltNonce) + 1).toString() },
      chains,
      knownSafeAddresses,
    )
  }

  return replayedSafe.saltNonce
}
