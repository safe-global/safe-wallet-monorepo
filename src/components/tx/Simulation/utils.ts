import { TENDERLY_SIMULATE_ENDPOINT_URL, TENDERLY_ORG_NAME, TENDERLY_PROJECT_NAME } from '@/config/constants'
import type { TenderlySimulatePayload, TenderlySimulation } from '@/components/tx/Simulation/types'

export const getSimulation = async (tx: TenderlySimulatePayload): Promise<TenderlySimulation> => {
  const data = await fetch(TENDERLY_SIMULATE_ENDPOINT_URL, {
    method: 'POST',
    body: JSON.stringify(tx),
  }).then((res) => res.json())

  return data as TenderlySimulation
}

export const getSimulationLink = (simulationId: string): string => {
  return `https://dashboard.tenderly.co/public/${TENDERLY_ORG_NAME}/${TENDERLY_PROJECT_NAME}/simulator/${simulationId}`
}
