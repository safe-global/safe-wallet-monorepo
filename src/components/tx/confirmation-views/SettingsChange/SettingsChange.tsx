import { Paper, Typography, Box, Divider, SvgIcon } from '@mui/material'
import EthHashInfo from '@/components/common/EthHashInfo'
import type { NarrowConfirmationViewProps } from '../types'
import { OwnerList } from '@/components/tx-flow/common/OwnerList'
import MinusIcon from '@/public/images/common/minus.svg'
import commonCss from '@/components/tx-flow/common/styles.module.css'
import useSafeInfo from '@/hooks/useSafeInfo'
import { SettingsInfoType, type SettingsChange } from '@safe-global/safe-gateway-typescript-sdk'

export interface SettingsChangeProps extends NarrowConfirmationViewProps {
  txInfo: SettingsChange
}

const SettingsChange: React.FC<SettingsChangeProps> = ({ txInfo }) => {
  const { safe } = useSafeInfo()
  const params = txInfo.settingsInfo

  if (!params || params.type === SettingsInfoType.REMOVE_OWNER) return null

  return (
    <>
      {'oldOwner' in params && (
        <Paper sx={{ backgroundColor: ({ palette }) => palette.warning.background, p: 2 }}>
          <Typography color="text.secondary" mb={2} display="flex" alignItems="center">
            <SvgIcon component={MinusIcon} inheritViewBox fontSize="small" sx={{ mr: 1 }} />
            Previous signer
          </Typography>
          <EthHashInfo
            name={params.oldOwner.name}
            address={params.oldOwner.value}
            shortAddress={false}
            showCopyButton
            hasExplorer
          />
        </Paper>
      )}
      {'owner' in params && <OwnerList owners={[params.owner]} />}
      {'newOwner' in params && <OwnerList owners={[params.newOwner]} />}
      {'threshold' in params && (
        <>
          <Divider className={commonCss.nestedDivider} />

          <Box>
            <Typography variant="body2">Any transaction requires the confirmation of:</Typography>
            <Typography>
              <b>{params.threshold}</b> out of <b>{safe.owners.length + ('removedOwner' in params ? 0 : 1)} signers</b>
            </Typography>
          </Box>
        </>
      )}
      <Divider className={commonCss.nestedDivider} />
    </>
  )
}

export default SettingsChange
