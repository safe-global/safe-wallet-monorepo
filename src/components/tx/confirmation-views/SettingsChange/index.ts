import SettingsChange from './SettingsChange'

export default SettingsChange
