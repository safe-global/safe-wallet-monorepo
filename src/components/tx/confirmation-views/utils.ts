import type { ChangeThreshold, SettingsChange, TransactionInfo } from '@safe-global/safe-gateway-typescript-sdk'
import { SettingsInfoType, TransactionInfoType } from '@safe-global/safe-gateway-typescript-sdk'
import { ConfirmBatchFlow } from '@/components/tx-flow/flows'

export const isSettingsChangeView = (txInfo: TransactionInfo) => txInfo.type === TransactionInfoType.SETTINGS_CHANGE

export const isConfirmBatchView = (txFlow?: JSX.Element) => txFlow?.type === ConfirmBatchFlow

export const isChangeThresholdView = (
  txInfo: TransactionInfo,
): txInfo is SettingsChange & { settingsInfo: ChangeThreshold } =>
  txInfo.type === TransactionInfoType.SETTINGS_CHANGE && txInfo.settingsInfo?.type === SettingsInfoType.CHANGE_THRESHOLD
