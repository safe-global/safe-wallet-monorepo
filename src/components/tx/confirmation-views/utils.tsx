import { TransactionInfoType } from '@safe-global/safe-gateway-typescript-sdk'
import type { NarrowConfirmationViewProps } from './types'
import type { SettingsChange } from '@safe-global/safe-gateway-typescript-sdk'
import SettingsChangeComponent from './SettingsChange'

export const getConfirmationViewComponent = (txType: TransactionInfoType, props: NarrowConfirmationViewProps) => {
  if (txType === TransactionInfoType.SETTINGS_CHANGE)
    return <SettingsChangeComponent txDetails={props.txDetails} txInfo={props.txInfo as SettingsChange} />

  return null
}
