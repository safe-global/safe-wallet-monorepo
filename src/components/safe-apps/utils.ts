import { BaseTransaction, ChainInfo } from '@gnosis.pm/safe-apps-sdk'
import { formatVisualAmount } from '@/utils/formatters'
import { validateAddress } from '@/utils/validation'
import { isHexString, toUtf8String } from 'ethers/lib/utils'

const validateTransaction = (t: BaseTransaction): boolean => {
  if (!['string', 'number'].includes(typeof t.value)) {
    return false
  }

  if (typeof t.value === 'string' && !/^(0x)?[0-9a-f]+$/i.test(t.value)) {
    return false
  }

  const isAddressValid = validateAddress(t.to) === undefined
  return isAddressValid && !!t.data && typeof t.data === 'string'
}

export const isTxValid = (txs: BaseTransaction[]) => txs.length && txs.every((t) => validateTransaction(t))

export const getInteractionTitle = (value?: string, chain?: ChainInfo) => {
  const { decimals, symbol } = chain!.nativeCurrency
  return `Interact with${
    Number(value) !== 0 ? ` (and send ${formatVisualAmount(value || 0, decimals)} ${symbol} to)` : ''
  }:`
}

export const convertToHumanReadableMessage = (message: string): string => {
  if (!isHexString(message)) {
    return message
  }

  return toUtf8String(message)
}
