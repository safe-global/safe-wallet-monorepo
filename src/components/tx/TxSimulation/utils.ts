import { generatePreValidatedSignature } from '@gnosis.pm/safe-core-sdk/dist/src/utils/signatures'
import EthSafeTransaction from '@gnosis.pm/safe-core-sdk/dist/src/utils/transactions/SafeTransaction'
import { encodeMultiSendData } from '@gnosis.pm/safe-core-sdk/dist/src/utils/transactions/utils'
import { FEATURES, type SafeInfo, type ChainInfo } from '@gnosis.pm/safe-react-gateway-sdk'
import type { MetaTransactionData, SafeTransaction } from '@gnosis.pm/safe-core-sdk-types'

import {
  getMultiSendCallOnlyContractInstance,
  getSpecificGnosisSafeContractInstance,
} from '@/services/contracts/safeContracts'
import { TENDERLY_SIMULATE_ENDPOINT_URL, TENDERLY_ORG_NAME, TENDERLY_PROJECT_NAME } from '@/config/constants'
import { hasFeature } from '@/utils/chains'
import type { StateObject, TenderlySimulatePayload, TenderlySimulation } from '@/components/tx/TxSimulation/types'

export const isTxSimulationEnabled = (chain?: ChainInfo): boolean => {
  if (!chain) {
    return false
  }

  const isSimulationEnvSet =
    Boolean(TENDERLY_SIMULATE_ENDPOINT_URL) && Boolean(TENDERLY_ORG_NAME) && Boolean(TENDERLY_PROJECT_NAME)

  return isSimulationEnvSet && hasFeature(chain, FEATURES.TX_SIMULATION)
}

export const getSimulation = async (tx: TenderlySimulatePayload): Promise<TenderlySimulation> => {
  const data = await fetch(TENDERLY_SIMULATE_ENDPOINT_URL, {
    method: 'POST',
    body: JSON.stringify(tx),
  }).then((res) => res.json())

  return data as TenderlySimulation
}

export const getSimulationLink = (simulationId: string): string => {
  return `https://dashboard.tenderly.co/public/${TENDERLY_ORG_NAME}/${TENDERLY_PROJECT_NAME}/simulator/${simulationId}`
}

type SingleTransactionSimulationParams = {
  safe: SafeInfo
  executionOwner: string
  transactions: SafeTransaction
  gasLimit: number
  canExecute: boolean
}

type MultiSendTransactionSimulationParams = {
  safe: SafeInfo
  executionOwner: string
  transactions: MetaTransactionData[]
  gasLimit: number
  canExecute: boolean
}

export type SimulationTxParams = SingleTransactionSimulationParams | MultiSendTransactionSimulationParams

export const _getSingleTransactionPayload = (
  params: SingleTransactionSimulationParams,
): Pick<TenderlySimulatePayload, 'to' | 'input'> => {
  // If a transaction is executable we simulate with the proposed/selected gasLimit and the actual signatures
  let transaction = params.transactions

  // Otherwise we overwrite the threshold to 1 on Tenderly and create a signature
  if (!params.canExecute) {
    const simulatedTransaction = new EthSafeTransaction(params.transactions.data)
    simulatedTransaction.addSignature(generatePreValidatedSignature(params.executionOwner))

    transaction = simulatedTransaction
  }

  const instance = getSpecificGnosisSafeContractInstance(params.safe)

  const input = instance.encode('execTransaction', [
    transaction.data.to,
    transaction.data.value,
    transaction.data.data,
    transaction.data.operation,
    transaction.data.safeTxGas,
    transaction.data.baseGas,
    transaction.data.gasPrice,
    transaction.data.gasToken,
    transaction.data.refundReceiver,
    transaction.encodedSignatures(),
  ])

  return {
    to: instance.getAddress(),
    input,
  }
}

export const _getMultiSendCallOnlyPayload = (
  params: MultiSendTransactionSimulationParams,
): Pick<TenderlySimulatePayload, 'to' | 'input'> => {
  const data = encodeMultiSendData(params.transactions)
  const instance = getMultiSendCallOnlyContractInstance(params.safe.chainId, params.safe.version)

  return {
    to: instance.getAddress(),
    input: instance.encode('multiSend', [data]),
  }
}

/* We need to overwrite the threshold stored in smart contract storage to 1
 to do a proper simulation that takes transaction guards into account.
 The threshold is stored in storage slot 4 and uses full 32 bytes slot
 Safe storage layout can be found here:
 https://github.com/gnosis/safe-contracts/blob/main/contracts/libraries/GnosisSafeStorage.sol */
const THRESHOLD_ONE_STORAGE_OVERRIDE = {
  [`0x${'4'.padStart(64, '0')}`]: `0x${'1'.padStart(64, '0')}`,
}

export const _getStateOverride = (
  address: string,
  balance?: string,
  code?: string,
  storage?: Record<string, string>,
): Record<string, StateObject> => {
  return {
    [address]: {
      balance,
      code,
      storage,
    },
  }
}

const isSingleTransactionSimulation = (params: SimulationTxParams): params is SingleTransactionSimulationParams => {
  return !Array.isArray(params.transactions)
}

export const getSimulationPayload = (params: SimulationTxParams): TenderlySimulatePayload => {
  const payload = isSingleTransactionSimulation(params)
    ? _getSingleTransactionPayload(params)
    : _getMultiSendCallOnlyPayload(params)

  return {
    ...payload,
    network_id: params.safe.chainId,
    from: params.executionOwner,
    gas: params.gasLimit,
    // With gas price 0 account don't need token for gas
    gas_price: '0',
    state_objects: _getStateOverride(params.safe.address.value, undefined, undefined, THRESHOLD_ONE_STORAGE_OVERRIDE),
    save: true,
    save_if_fails: true,
  }
}
