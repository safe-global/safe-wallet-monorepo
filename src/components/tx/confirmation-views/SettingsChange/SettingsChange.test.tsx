import { render } from '@/tests/test-utils'
import type { TransactionDetails } from '@safe-global/safe-gateway-typescript-sdk'
import { SettingsInfoType } from '@safe-global/safe-gateway-typescript-sdk'
import SettingsChange from './index'
import { ownerAddress, txInfo } from './utils'

describe('SettingsChange', () => {
  it('should display the SettingsChange component with owner details', () => {
    const { container, getByText } = render(<SettingsChange txDetails={{} as TransactionDetails} txInfo={txInfo} />)

    expect(container).toMatchSnapshot()
    expect(getByText('New signer')).toBeInTheDocument()
    expect(getByText(ownerAddress)).toBeInTheDocument()
  })

  it('should display the SettingsChange component with newOwner details', () => {
    const newOwnerAddress = '0x0000000000000000'

    const { container, getByText } = render(
      <SettingsChange
        txDetails={{} as TransactionDetails}
        txInfo={{
          ...txInfo,
          settingsInfo: {
            type: SettingsInfoType.SWAP_OWNER,
            oldOwner: {
              value: ownerAddress,
              name: 'Bob',
              logoUri: 'http://bob.com',
            },
            newOwner: {
              value: newOwnerAddress,
              name: 'Alice',
              logoUri: 'http://alice.com',
            },
          },
        }}
      />,
    )

    expect(container).toMatchSnapshot()
    expect(getByText('Previous signer')).toBeInTheDocument()
    expect(getByText(newOwnerAddress)).toBeInTheDocument()
  })
})
