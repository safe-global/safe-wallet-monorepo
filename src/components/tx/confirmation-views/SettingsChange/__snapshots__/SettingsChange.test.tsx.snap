// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsChange should display the SettingsChange component with newOwner details 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-1f23gi7-MuiPaper-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-ya5m4n-MuiTypography-root"
    >
      <mock-icon
        aria-hidden="true"
        classname="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-gxmw0o-MuiSvgIcon-root"
        focusable="false"
      />
      Previous signer
    </p>
    <div
      class="container"
    >
      <div
        class="avatarContainer"
        style="width: 40px; height: 40px;"
      >
        <div
          class="icon"
          style="background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA4IDgiIHNoYXBlLXJlbmRlcmluZz0ib3B0aW1pemVTcGVlZCIgd2lkdGg9IjY0IiBoZWlnaHQ9IjY0Ij48cGF0aCBmaWxsPSJoc2woMzI1IDQ4JSAyMiUpIiBkPSJNMCwwSDhWOEgweiIvPjxwYXRoIGZpbGw9ImhzbCgyNDAgOTklIDQ5JSkiIGQ9Ik0xLDBoMXYxaC0xek02LDBoMXYxaC0xek0yLDBoMXYxaC0xek01LDBoMXYxaC0xek0zLDBoMXYxaC0xek00LDBoMXYxaC0xek0wLDFoMXYxaC0xek03LDFoMXYxaC0xek0wLDJoMXYxaC0xek03LDJoMXYxaC0xek0xLDJoMXYxaC0xek02LDJoMXYxaC0xek0yLDJoMXYxaC0xek01LDJoMXYxaC0xek0xLDNoMXYxaC0xek02LDNoMXYxaC0xek0xLDRoMXYxaC0xek02LDRoMXYxaC0xek0xLDVoMXYxaC0xek02LDVoMXYxaC0xek0yLDVoMXYxaC0xek01LDVoMXYxaC0xek0yLDdoMXYxaC0xek01LDdoMXYxaC0xek0zLDdoMXYxaC0xek00LDdoMXYxaC0xeiIvPjxwYXRoIGZpbGw9ImhzbCgxNTUgOTUlIDU5JSkiIGQ9Ik0xLDFoMXYxaC0xek02LDFoMXYxaC0xek0zLDFoMXYxaC0xek00LDFoMXYxaC0xek0zLDJoMXYxaC0xek00LDJoMXYxaC0xek0xLDdoMXYxaC0xek02LDdoMXYxaC0xeiIvPjwvc3ZnPg==); width: 40px; height: 40px;"
        />
      </div>
      <div
        class="MuiBox-root css-1lchl8k"
      >
        <div
          class="MuiBox-root css-171onha"
          title="Bob"
        >
          <div
            class="MuiBox-root css-1o4wo1x"
          >
            Bob
          </div>
        </div>
        <div
          class="addressContainer"
        >
          <div
            class="MuiBox-root css-b5p5gz"
          >
            <span
              aria-label="Copy to clipboard"
              class=""
              data-mui-internal-clone-element="true"
              style="cursor: pointer;"
            >
              <span>
                0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045
              </span>
            </span>
          </div>
          <span
            aria-label="Copy to clipboard"
            class=""
            data-mui-internal-clone-element="true"
            style="cursor: pointer;"
          >
            <button
              aria-label="Copy to clipboard"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-nmr6e8-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <mock-icon
                aria-hidden="true"
                classname="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-j1da6x-MuiSvgIcon-root"
                data-testid="copy-btn-icon"
                focusable="false"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </span>
          <div
            class="MuiBox-root css-yjghm1"
          >
            <button
              aria-label=""
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1o2vr87-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              data-testid="explorer-btn"
              href=""
              rel="noreferrer"
              tabindex="0"
              target="_blank"
              type="button"
            >
              <mock-icon
                aria-hidden="true"
                classname="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-11dtfly-MuiSvgIcon-root"
                focusable="false"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 container css-1bq3jbz-MuiPaper-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-1ab5q2h-MuiTypography-root"
    >
      <mock-icon
        aria-hidden="true"
        classname="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-gxmw0o-MuiSvgIcon-root"
        focusable="false"
      />
      New signer
    </p>
    <div
      class="container"
    >
      <div
        class="avatarContainer"
        style="width: 32px; height: 32px;"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-1sj56zx-MuiSkeleton-root"
          style="width: 32px; height: 32px;"
        />
      </div>
      <div
        class="MuiBox-root css-1lchl8k"
      >
        <div
          class="MuiBox-root css-171onha"
          title="Alice"
        >
          <div
            class="MuiBox-root css-1o4wo1x"
          >
            Alice
          </div>
        </div>
        <div
          class="addressContainer"
        >
          <div
            class="MuiBox-root css-b5p5gz"
          >
            <span
              aria-label="Copy to clipboard"
              class=""
              data-mui-internal-clone-element="true"
              style="cursor: pointer;"
            >
              <span>
                0x0000000000000000
              </span>
            </span>
          </div>
          <span
            aria-label="Copy to clipboard"
            class=""
            data-mui-internal-clone-element="true"
            style="cursor: pointer;"
          >
            <button
              aria-label="Copy to clipboard"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-nmr6e8-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <mock-icon
                aria-hidden="true"
                classname="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-j1da6x-MuiSvgIcon-root"
                data-testid="copy-btn-icon"
                focusable="false"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </span>
          <div
            class="MuiBox-root css-yjghm1"
          >
            <button
              aria-label=""
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1o2vr87-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              data-testid="explorer-btn"
              href=""
              rel="noreferrer"
              tabindex="0"
              target="_blank"
              type="button"
            >
              <mock-icon
                aria-hidden="true"
                classname="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-11dtfly-MuiSvgIcon-root"
                focusable="false"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth nestedDivider css-da8pia-MuiDivider-root"
  />
</div>
`;

exports[`SettingsChange should display the SettingsChange component with owner details 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 container css-1bq3jbz-MuiPaper-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-1ab5q2h-MuiTypography-root"
    >
      <mock-icon
        aria-hidden="true"
        classname="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-gxmw0o-MuiSvgIcon-root"
        focusable="false"
      />
      New signer
    </p>
    <div
      class="container"
    >
      <div
        class="avatarContainer"
        style="width: 32px; height: 32px;"
      >
        <div
          class="icon"
          style="background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA4IDgiIHNoYXBlLXJlbmRlcmluZz0ib3B0aW1pemVTcGVlZCIgd2lkdGg9IjY0IiBoZWlnaHQ9IjY0Ij48cGF0aCBmaWxsPSJoc2woMzI1IDQ4JSAyMiUpIiBkPSJNMCwwSDhWOEgweiIvPjxwYXRoIGZpbGw9ImhzbCgyNDAgOTklIDQ5JSkiIGQ9Ik0xLDBoMXYxaC0xek02LDBoMXYxaC0xek0yLDBoMXYxaC0xek01LDBoMXYxaC0xek0zLDBoMXYxaC0xek00LDBoMXYxaC0xek0wLDFoMXYxaC0xek03LDFoMXYxaC0xek0wLDJoMXYxaC0xek03LDJoMXYxaC0xek0xLDJoMXYxaC0xek02LDJoMXYxaC0xek0yLDJoMXYxaC0xek01LDJoMXYxaC0xek0xLDNoMXYxaC0xek02LDNoMXYxaC0xek0xLDRoMXYxaC0xek02LDRoMXYxaC0xek0xLDVoMXYxaC0xek02LDVoMXYxaC0xek0yLDVoMXYxaC0xek01LDVoMXYxaC0xek0yLDdoMXYxaC0xek01LDdoMXYxaC0xek0zLDdoMXYxaC0xek00LDdoMXYxaC0xeiIvPjxwYXRoIGZpbGw9ImhzbCgxNTUgOTUlIDU5JSkiIGQ9Ik0xLDFoMXYxaC0xek02LDFoMXYxaC0xek0zLDFoMXYxaC0xek00LDFoMXYxaC0xek0zLDJoMXYxaC0xek00LDJoMXYxaC0xek0xLDdoMXYxaC0xek02LDdoMXYxaC0xeiIvPjwvc3ZnPg==); width: 32px; height: 32px;"
        />
      </div>
      <div
        class="MuiBox-root css-1lchl8k"
      >
        <div
          class="MuiBox-root css-171onha"
          title="Nevinha"
        >
          <div
            class="MuiBox-root css-1o4wo1x"
          >
            Nevinha
          </div>
        </div>
        <div
          class="addressContainer"
        >
          <div
            class="MuiBox-root css-b5p5gz"
          >
            <span
              aria-label="Copy to clipboard"
              class=""
              data-mui-internal-clone-element="true"
              style="cursor: pointer;"
            >
              <span>
                0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045
              </span>
            </span>
          </div>
          <span
            aria-label="Copy to clipboard"
            class=""
            data-mui-internal-clone-element="true"
            style="cursor: pointer;"
          >
            <button
              aria-label="Copy to clipboard"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-nmr6e8-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <mock-icon
                aria-hidden="true"
                classname="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-j1da6x-MuiSvgIcon-root"
                data-testid="copy-btn-icon"
                focusable="false"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </span>
          <div
            class="MuiBox-root css-yjghm1"
          >
            <button
              aria-label=""
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1o2vr87-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              data-testid="explorer-btn"
              href=""
              rel="noreferrer"
              tabindex="0"
              target="_blank"
              type="button"
            >
              <mock-icon
                aria-hidden="true"
                classname="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-11dtfly-MuiSvgIcon-root"
                focusable="false"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth nestedDivider css-da8pia-MuiDivider-root"
  />
  <div
    class="MuiBox-root css-0"
  >
    <p
      class="MuiTypography-root MuiTypography-body2 css-ceebrx-MuiTypography-root"
    >
      Any transaction requires the confirmation of:
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-1pqjor9-MuiTypography-root"
    >
      <b>
        1
      </b>
       out of
       
      <b>
        1
         signers
      </b>
    </p>
  </div>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth nestedDivider css-da8pia-MuiDivider-root"
  />
</div>
`;
